gradio/routes.py

```diff
@@ -235,6 +236,72 @@ def toorjson(value):
 file_upload_statuses = FileUploadProgress()
 
 
+@dataclass
+class NodeProxyCacheEntry:
+    head: bytearray
+    subs: list[asyncio.Queue[bytes | None]]
+    resp: asyncio.Future[httpx.Response | None]
+
+
+class NodeProxyCache:
```

Comment:
I would add a docstring here explaining how this works at a high level. Its really neat!